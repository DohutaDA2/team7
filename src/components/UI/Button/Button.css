.Button {
  display: block;
  background-color: transparent;
  border: none;
  color: #ca752d;
  outline: none;
  font-family: Arial;
  padding: 10px;
  font-weight: bold;
  font-size: 10;
  margin: auto;
  margin-top: 10px;
  width: -webkit-fill-available;
  border-radius: 4px;
  -webkit-transition: all 0.3s ease-in-out;
  min-width: 100px;
}

.Button:disabled {
  color: #7f7f7f;
  background-color: #f1f1f1;
  cursor: not-allowed;
  -webkit-transition: all 0.3s ease-in-out;
}

.PrimaryOrange {
  background: #fff;
  border: solid 1px #fff;
  color: #d65219;
  box-shadow: 0px 4px 10px -2px rgb(0, 0, 0, 0.7);
  -webkit-transition: all 0.3s ease-in-out;
}

.SecondaryOrange {
  border: solid 1px #fff;
  background-color: #d65219;
  color: #fff;
  -webkit-transition: all 0.3s ease-in-out;
}

.PrimaryGreen {
  border: solid 1px #559900;
  background: #559900;
  color: #fff;
  box-shadow: 0px 3px 5px 0px #c2c2c2;
  -webkit-transition: all 0.3s ease-in-out;
}

.SecondaryGreen {
  border: solid 1px;
  background-color: #fff;
  color: #559900;
  -webkit-transition: all 0.3s ease-in-out;
}

.Button:hover {
  background-color: #f3cdad;
  -webkit-transition: all 0.3s ease-in-out;
}

.SecondaryGreen:hover,
.PrimaryGreen:hover {
  background-color: #76ad32;
  border: solid 1px #76ad32;
  color: #fff;
  -webkit-transition: all 0.3s ease-in-out;
}

.SecondaryOrange:hover,
.PrimaryOrange:hover {
  background-color: #e18233;
  border: solid 1px #fff;
  color: #fff;
  -webkit-transition: all 0.3s ease-in-out;
}

.OK {
  display: block;
  margin: auto;
  border: solid 1px #3b5998;
  background-color: #3b5998;
  color: #ffffff;
  -webkit-transition: all 0.3s ease-in-out;
}

.OK:hover {
  border: solid 1px #3b5998;
  background-color: #5772aa;
  color: #ffffff;
  -webkit-transition: all 0.3s ease-in-out;
}

.Cancel {
  display: block;
  margin: auto;
  border: solid 1px #3b5998;
  background-color: #ffffff;
  color: #3b5998;
  -webkit-transition: all 0.3s ease-in-out;
}

.Cancel:hover {
  border: solid 1px #3b5998;
  background-color: #5772aa;
  color: #ffffff;
  -webkit-transition: all 0.3s ease-in-out;
}

.Action {
  max-width: 100px;
  box-sizing: border-box;
  color: rgb(0, 0, 0, 1);
  font-size: 12px;
  font-weight: 300;
  margin: 0;
  border-radius: 6px;
}

.Action:hover {
  background-color: rgb(0, 0, 0, 0.05);
}

@media screen and (min-width: 960px) {
  .Button {
    max-width: 600px;
    margin-top: 20px;
  }

  .Action {
    max-width: 70px;
    margin: 0;
  }
}

@media screen and (max-width: 960px) {
  .Button {
    max-width: 300px;
    margin-top: 10px;
  }

  .Action {
    max-width: 40px;
    margin: 0;
  }
}

@media screen and (max-width: 340px) {
  .Button {
    width: 200px;
    margin-top: 10px;
  }

  .Action {
    width: 40px;
    margin: 0;
  }
}
