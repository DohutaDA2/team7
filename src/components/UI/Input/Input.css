.Input {
  width: 100%;
  padding: 5px 0px;
  box-sizing: border-box;
  -webkit-transition: all 0.3s ease-in-out;
}

.Label {
  display: block;
  margin-bottom: 4px;
  margin-top: 10px;
}

.LabelEmail {
  font-family: Arial;
  font-size: 10;
  color: blue;
}
.LabelPassword {
  font-family: Arial;
  font-size: 11;
  color: grey;
}

.InputElement {
  max-width: 400px;
  outline: none;
  border: 1px solid #ccc;
  background-color: white;
  font: inherit;
  padding: 6px 20px;
  margin: auto;
  display: block;
  width: 100%;
  box-sizing: border-box;
  border-radius: 24px;
  -webkit-transition: all 0.3s ease-in-out;
}

.InputElement:focus {
  outline: none;
  box-shadow: 0px 3px 8px 0px #c2c2c2;
  -webkit-transition: all 0.3s ease-in-out;
}

.Invalid {
  border: 1px solid rgba(255, 0, 0, 0.863);
  background-color: rgb(255, 152, 152);
  font-weight: 600;
}

.Email::-placeholder {
  color: rgb(0, 148, 62);
  opacity: 1; /* Firefox */
}

.DatePicker {
  max-width: 200px;
  outline: none;
  border: 1px solid #ccc;
  background-color: white;
  font: inherit;
  padding: 6px 20px;
  margin: auto;
  display: block;
  width: 100%;
  border-radius: 24px;
  -webkit-transition: all 0.3s ease-in-out;
  text-align: center;
  cursor: pointer;
  text-align: center;
  color: transparent;
  text-shadow: 0 0 0 #000000;
  &:focus {
    outline: none;
  }
}

.Input:disabled {
  background-color: #ccc;
  cursor: not-allowed;
}
