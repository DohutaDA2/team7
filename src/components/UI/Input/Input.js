import React from "react";
import classes from "./Input.css";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker-cssmodules.css";
import moment from "moment";

const input = props => {
  let inputElement = null;
  let label = null;
  const inputClasses = [classes.InputElement];

  if (props.status && props.shouldValidate && props.touched) {
    inputClasses.push(classes.Invalid);
  }

  if (props.elementConfig.type === "email") {
    inputClasses.push(classes.Email);
  }

  if (props.label === "Email") {
    label = (
      <label className={[classes.Label, classes.LabelEmail].join(" ")}>
        {props.label}
      </label>
    );
  } else if (props.label === "Mật khẩu") {
    label = (
      <label className={[classes.Label, classes.LabelPassword].join(" ")}>
        {props.label}
      </label>
    );
  } else {
    label = (
      <label
        style={{
          display: "block",
          textAlign: "center",
          paddingBottom: "5px",
          color: "#a1a1a1"
        }}
      >
        {props.elementConfig.placeholder}
      </label>
    );
  }

  switch (props.elementType) {
    case "input":
      inputElement = (
        <input
          className={inputClasses.join(" ")}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
          disabled={props.disabled}
        />
      );
      break;
    case "textarea":
      inputElement = (
        <textarea
          className={inputClasses.join(" ")}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
        />
      );
      break;
    case "select":
      inputElement = (
        <select
          className={inputClasses.join(" ")}
          value={props.value}
          onChange={props.changed}
        >
          {props.elementConfig.options.map(option => (
            <option key={option.value} value={option.value}>
              {option.displayValue}
            </option>
          ))}
        </select>
      );
      break;
    case "datepicker":
      inputElement = (
        <div style={{ display: "block", margin: "auto", maxWidth: "245px" }}>
          <DatePicker
            style={classes.DatePicker}
            className={classes.DatePicker}
            {...props.elementConfig}
            selected={props.value}
            onChange={props.changed}
            minDate={moment().subtract(24, "months")}
            maxDate={moment()}
          />
        </div>
      );
      break;
    default:
      inputElement = (
        <input
          className={inputClasses.join(" ")}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
          disabled={props.disabled}
        />
      );
  }

  return (
    <div className={classes.Input}>
      {label}
      {inputElement}
    </div>
  );
};

export default input;
