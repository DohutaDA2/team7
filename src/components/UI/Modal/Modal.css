.Modal {
  position: fixed;
  z-index: 200;
  background-color: #f1f1f1;
  width: 70%;
  border: none;
  box-shadow: 0px 12px 24px rgb(0, 0, 0, 0.4);
  padding: 16px;
  left: 15%;
  top: 30%;
  box-sizing: border-box;
  border-radius: 10px;
  transition: all 0.5s ease-in-out;
}

@media (min-width: 600px) {
  .Modal {
    width: 500px;
    left: calc(50% - 250px);
  }
}
