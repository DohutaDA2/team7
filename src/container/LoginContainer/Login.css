.BoxLogin{
  display: block;
  margin: auto;
  padding: 20px;
  position: relative;
  max-width: 300px;
}

.LoginTitle{
  text-align: center;
  font-weight: 300;
  font-size: 20px;
}

.ModalIcon{
  width: 40px;
  height: 40px;
  display: block;
  margin: auto;
  padding-bottom: 40px;
  text-align: center;
  color: red;
  font-size: -webkit-xxx-large;
}

.ModalMessage{
  font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
  text-align: center;
  font-size: 24px;
  font-weight: 300;
  margin: 20px auto 20px auto;
}
