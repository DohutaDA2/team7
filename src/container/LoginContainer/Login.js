import React, { Component } from "react";
import Spinner from "../../components/UI/Spinner/Spinner";
import Modal from "../../components/UI/Modal/Modal";
import Layout from "../../hoc/Layout/Layout";
import Input from "../../components/UI/Input/Input";
import Button from "../../components/UI/Button/Button";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import classes from "./Login.css";
import { checkValidityLogin } from "../../hoc/business/checkValidity";
import { getAuth } from "../../hoc/business/loginHandler";

class Login extends Component {
  state = {
    loading: false,
    isSignUp: false,
    errorMessage: [],
    controls: {
      email: {
        elementType: "input",
        elementConfig: {
          type: "email",
          placeholder: "Email"
        },
        value: "",
        validation: {
          required: true,
          isEmail: true
        },
        status: {
          valid: false,
          errorMessage: ""
        },
        touched: false
      },
      password: {
        elementType: "input",
        elementConfig: {
          type: "password",
          placeholder: "Mật khẩu"
        },
        value: "",
        validation: {
          required: true,
          minLength: 8,
          isRequireLetterNumberSymbol: true
        },
        status: {
          valid: false,
          errorMessage: ""
        },
        touched: false
      }
    }
  };

  inputChangedHandler = (event, controlName) => {
    const updatedControls = {
      ...this.state.controls,
      [controlName]: {
        ...this.state.controls[controlName],
        value: event.target.value,
        status: checkValidityLogin(
          controlName,
          event.target.value,
          this.state.controls[controlName].validation
        ),
        touched: true
      }
    };
    this.setState({ controls: updatedControls });
  };

  /**
   * Signup process handler
   */
  // signUpHandler = event => {
  //   event.preventDefault();
  //   let errorMessage = [];
  //   if (
  //     this.state.controls.email.value === "" &&
  //     this.state.controls.password.value === ""
  //   )
  //     errorMessage.push("EMAIL và PASSWORD không được rỗng");
  //   else {
  //     if (this.state.controls.email.value === "")
  //       errorMessage.push("EMAIL không được rỗng");
  //     if (this.state.controls.password.value === "")
  //       errorMessage.push("PASSWORD không được rỗng");
  //   }
  //   for (let key in this.state.controls) {
  //     if (this.state.controls[key].status.errorMessage !== "") {
  //       errorMessage = errorMessage.concat(
  //         this.state.controls[key].status.errorMessage
  //       );
  //     }
  //   }
  //   this.setState({ errorMessage: errorMessage });

  //   if (errorMessage.length === 0) {
  //     this.setState({ loading: true });
  //     const requestData = {
  //       email: this.state.controls.email.value,
  //       password: this.state.controls.password.value,
  //       returnSecureToken: true
  //     };
  //     callServer("su", requestData)
  //       .then(res => {
  //         this.setState({ loading: false });
  //         this.signIn(res);
  //       })
  //       .catch(err => {
  //         // console.log(err);
  //         errorMessage.push("Mã lỗi: " + err.code);
  //         errorMessage.push(err.message);
  //         this.setState({ loading: false, errorMessage: errorMessage });
  //       });
  //   }
  // };

  /**
   * Signin process handler
   */
  // signInHandler = event => {
  //   event.preventDefault();
  //   let errorMessage = [];
  //   if (
  //     this.state.controls.email.value === "" &&
  //     this.state.controls.password.value === ""
  //   )
  //     errorMessage.push("EMAIL và PASSWORD không được rỗng");
  //   else {
  //     if (this.state.controls.email.value === "")
  //       errorMessage.push("EMAIL không được rỗng");
  //     if (this.state.controls.password.value === "")
  //       errorMessage.push("PASSWORD không được rỗng");
  //   }
  //   for (let key in this.state.controls) {
  //     if (this.state.controls[key].status.errorMessage !== "") {
  //       errorMessage = errorMessage.concat(
  //         this.state.controls[key].status.errorMessage
  //       );
  //     }
  //   }
  //   this.setState({ errorMessage: errorMessage });

  //   if (errorMessage.length === 0) {
  //     this.setState({ loading: true });
  //     const requestData = {
  //       email: this.state.controls.email.value,
  //       password: this.state.controls.password.value,
  //       returnSecureToken: true
  //     };
  //     callServer("si", requestData)
  //       .then(res => {
  //         this.setState({ loading: false });
  //         this.signIn(res);
  //       })
  //       .catch(err => {
  //         // console.log(err);
  //         errorMessage.push("Mã lỗi: " + err.code);
  //         errorMessage.push(err.message);
  //         this.setState({ loading: false, errorMessage: errorMessage });
  //       });
  //   }
  // };

  /**
   * Signin process handler
   */
  signInHandler = event => {
    event.preventDefault();
    let errorMessage = [];
    if (
      this.state.controls.email.value === "" &&
      this.state.controls.password.value === ""
    )
      errorMessage.push("EMAIL và PASSWORD không được rỗng");
    else {
      if (this.state.controls.email.value === "")
        errorMessage.push("EMAIL không được rỗng");
      if (this.state.controls.password.value === "")
        errorMessage.push("PASSWORD không được rỗng");
    }
    for (let key in this.state.controls) {
      if (this.state.controls[key].status.errorMessage !== "") {
        errorMessage = errorMessage.concat(
          this.state.controls[key].status.errorMessage
        );
      }
    }
    this.setState({ errorMessage: errorMessage });

    if (errorMessage.length === 0) {
      this.setState({ loading: true });
      const requestData = {
        email: this.state.controls.email.value,
        password: this.state.controls.password.value
      };
      getAuth("si", requestData)
        .then(res => {
          this.setState({ loading: false });
          this.signIn(res);
        })
        .catch(err => {
          // console.log(err);
          errorMessage.push("Mã lỗi: " + err.code);
          errorMessage.push(err.message);
          this.setState({ loading: false, errorMessage: errorMessage });
        });
    }
  };

  
  /**
   * Signup process handler
   */
  signUpHandler = event => {
    event.preventDefault();
    let errorMessage = [];
    if (
      this.state.controls.email.value === "" &&
      this.state.controls.password.value === ""
    )
      errorMessage.push("EMAIL và PASSWORD không được rỗng");
    else {
      if (this.state.controls.email.value === "")
        errorMessage.push("EMAIL không được rỗng");
      if (this.state.controls.password.value === "")
        errorMessage.push("PASSWORD không được rỗng");
    }
    for (let key in this.state.controls) {
      if (this.state.controls[key].status.errorMessage !== "") {
        errorMessage = errorMessage.concat(
          this.state.controls[key].status.errorMessage
        );
      }
    }
    this.setState({ errorMessage: errorMessage });

    if (errorMessage.length === 0) {
      this.setState({ loading: true });
      const requestData = {
        email: this.state.controls.email.value,
        password: this.state.controls.password.value
      };
      getAuth("su", requestData)
        .then(res => {
          this.setState({ loading: false });
          this.signIn(res);
        })
        .catch(err => {
          // console.log(err);
          errorMessage.push("Mã lỗi: " + err.code);
          errorMessage.push(err.message);
          this.setState({ loading: false, errorMessage: errorMessage });
        });
    }
  };

  signIn = data => {
    const queryParams = [];
    for (let key in data) {
      queryParams.push(
        encodeURIComponent(key) + "=" + encodeURIComponent(data[key])
      );
    }
    const queryString = queryParams.join("&");
    this.props.history.push({
      pathname: "/home",
      search: "?" + queryString
    });
  };

  closeModalHanlder = () => {
    this.setState({ errorMessage: [] });
  };

  render() {
    let spinner = null;
    if (this.state.loading) {
      spinner = <Spinner />;
    }
    const formElementsArray = [];
    for (let key in this.state.controls) {
      formElementsArray.push({
        id: key,
        config: this.state.controls[key]
      });
    }
    const form = formElementsArray.map(formElement => (
      <Input
        key={formElement.id}
        label={
          formElement.config.elementConfig.placeholder
            ? formElement.config.elementConfig.placeholder
            : ""
        }
        elementType={formElement.config.elementType}
        elementConfig={formElement.config.elementConfig}
        value={formElement.config.value}
        status={!formElement.config.status.isValid}
        shouldValidate={formElement.config.validation}
        touched={formElement.config.touched}
        changed={event => this.inputChangedHandler(event, formElement.id)}
      />
    ));

    return (
      <div>
        {spinner}
        <Modal
          show={this.state.errorMessage.length !== 0}
          modalClosed={this.closeModalHanlder}
        >
          <div className={classes.ModalIcon}>
            <FontAwesomeIcon icon={faInfoCircle} />
          </div>
          {this.state.errorMessage.map((message, index) => (
            <p key={index} className={classes.ModalMessage}>
              {message}
            </p>
          ))}
          <Button btnType="OK" clicked={this.closeModalHanlder}>
            OK
          </Button>
        </Modal>
        <div style={{ backgroundColor: "#f7f7f7", height: "100vh" }}>
          <Layout
            styles={{ borderRadius: "4px", position: "relative", top: "10%" }}
          >
            <div className={classes.BoxLogin}>
              <p className={classes.LoginTitle}>Đăng ký bằng Email</p>
              {form}
              <br />
              <Button
                css={{ maxWidth: "300px" }}
                btnType="PrimaryGreen"
                clicked={event => this.signUpHandler(event)}
              >
                ĐĂNG KÝ
              </Button>
              <Button
                css={{ maxWidth: "300px" }}
                btnType="SecondaryGreen"
                clicked={event => this.signInHandler(event)}
              >
                ĐĂNG NHẬP
              </Button>
            </div>
          </Layout>
        </div>
      </div>
    );
  }
}

export default Login;
