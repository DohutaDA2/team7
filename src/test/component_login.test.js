import React from "react";
import { mount, shallow, render } from "enzyme";

import Login from "../container/LoginContainer/Login";
import Button from "../components/UI/Button/Button";
import Input from "../components/UI/Input/Input";
import Modal from "../components/UI/Modal/Modal";

describe("Kiểm tra component", () => {
  it("phải render chính xác và không render các component con của nó", () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });

  it("phải thay đổi state khi inputs thay đổi", () => {
    const component = mount(<Login />);

    component
      .find('input')
      .at(1)
      .simulate("change", { target: { value: "abcabcabc" } });
    expect(component.state().controls.email.status.isValid).toBe(false);

    component
      .find("input")
      .at(3)
      .simulate("change", { target: { value: "abcabcabc" } });
    expect(component.state().controls.password.status.isValid).toBe(false);

    component
      .find(Button)
      .at(2)
      .simulate("click")
      .simulate("click");

    expect(component.state().errorMessage.length).toBeGreaterThan(0);
    expect(component.state().controls.email.touched).toEqual(true);
    expect(component.state().controls.password.touched).toEqual(true);
    component.unmount();
  });
});
