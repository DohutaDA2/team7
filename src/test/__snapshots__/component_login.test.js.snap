// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kiểm tra component phải render chính xác và không render các component con của nó 1`] = `
<div>
  <Modal
    modalClosed={[Function]}
    show={false}
  >
    <div>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f05a",
              "M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z",
            ],
            "iconName": "info-circle",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </div>
    <button
      btnType="OK"
      clicked={[Function]}
    >
      OK
    </button>
  </Modal>
  <div
    style={
      Object {
        "backgroundColor": "#f7f7f7",
        "height": "100vh",
      }
    }
  >
    <Layout
      styles={
        Object {
          "borderRadius": "4px",
          "position": "relative",
          "top": "10%",
        }
      }
    >
      <div>
        <p>
          Đăng ký bằng Email
        </p>
        <input
          changed={[Function]}
          elementConfig={
            Object {
              "placeholder": "Email",
              "type": "email",
            }
          }
          elementType="input"
          key="email"
          label="Email"
          shouldValidate={
            Object {
              "isEmail": true,
              "required": true,
            }
          }
          status={true}
          touched={false}
          value=""
        />
        <input
          changed={[Function]}
          elementConfig={
            Object {
              "placeholder": "Mật khẩu",
              "type": "password",
            }
          }
          elementType="input"
          key="password"
          label="Mật khẩu"
          shouldValidate={
            Object {
              "isRequireLetterNumberSymbol": true,
              "minLength": 8,
              "required": true,
            }
          }
          status={true}
          touched={false}
          value=""
        />
        <br />
        <button
          btnType="PrimaryGreen"
          clicked={[Function]}
          css={
            Object {
              "maxWidth": "300px",
            }
          }
        >
          ĐĂNG KÝ
        </button>
        <button
          btnType="SecondaryGreen"
          clicked={[Function]}
          css={
            Object {
              "maxWidth": "300px",
            }
          }
        >
          ĐĂNG NHẬP
        </button>
      </div>
    </Layout>
  </div>
</div>
`;
